src/Compilers/CSharp/Portable/Utilities/InterceptableLocation.cs

```diff
@@ -0,0 +1,125 @@
+}
+
+#pragma warning disable RSEXPERIMENTAL002 // internal usage of experimental API
+/// <summary>
+/// Version 1 of the InterceptableLocation encoding.
+/// </summary>
+internal sealed class InterceptableLocation1 : InterceptableLocation
+{
+    private readonly ImmutableArray<byte> _checksum;
+    private readonly string _path;
+    private readonly int _position;
+    private readonly int _lineNumberOneIndexed;
+    private readonly int _characterNumberOneIndexed;
+    private string? _data;
+
+    internal InterceptableLocation1(ImmutableArray<byte> checksum, string path, int position, int lineNumberOneIndexed, int characterNumberOneIndexed)
+    {
+        if (checksum.Length != 16)
+        {
+            throw new ArgumentException(message: "checksum must be exactly 16 bytes in length", paramName: nameof(checksum));
+        }
+
+        _checksum = checksum;
+        _path = path;
+        _position = position;
+        _lineNumberOneIndexed = lineNumberOneIndexed;
+        _characterNumberOneIndexed = characterNumberOneIndexed;
+    }
+
+    public override string GetDisplayLocation()
+    {
+        // e.g. `C:\project\src\Program.cs(12,34)`
+        return $"{_path}({_lineNumberOneIndexed},{_characterNumberOneIndexed})";
+    }
+
+    public override int Version => 1;
+    public override string Data
+    {
+        get
+        {
+            if (_data is null)
+                _data = makeData();
+
+            return _data;
+
+            string makeData()
+            {
+                var builder = PooledBlobBuilder.GetInstance();
+                builder.WriteBytes(_checksum, start: 0, 16);
+                builder.WriteInt32(_position);
+
+                var displayFileName = Path.GetFileName(_path);
+                builder.WriteUTF8(displayFileName);
+
+                var bytes = builder.ToArray();
+                builder.Free();
+                return Convert.ToBase64String(bytes);
+            }
+        }
+    }
+
+    // Note: the goal of implementing equality here is so that incremental state tables etc. can detect and use it.
+    // This encoding which uses the checksum of the referenced file may not be stable across incremental runs in practice, but it seems correct in principle to implement equality here anyway.
+    public override bool Equals(object? obj)
+    {
+        if ((object)this == obj)
+            return true;
+
+        return obj is InterceptableLocation1 other
+            && _checksum.SequenceEqual(other._checksum)
+            && _path == other._path
+            && _position == other._position
+            && _lineNumberOneIndexed == other._lineNumberOneIndexed
+            && _characterNumberOneIndexed == other._characterNumberOneIndexed;
+    }
+
+    public override int GetHashCode()
+    {
+        return Hash.Combine(
+           BinaryPrimitives.ReadInt32LittleEndian(_checksum.AsSpan()),
+           Hash.Combine(
+               _path.GetHashCode(),
```

Comment:
nit: we may be able to skip hashing the `_path` without reducing hash quality (content hash + line + character seems pretty good already)

Reply:
`_path` only gets its value from `SyntaxTree.FilePath` which is documented to never return null.